include/exadg/postprocessor/pointwise_output_generator_base.h

```diff
@@ -0,0 +1,198 @@
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *  ______________________________________________________________________
+ */
+
+#ifndef INCLUDE_COMPRESSIBLE_NAVIER_STOKES_POSTPROCESSOR_POINTWISE_OUTPUT_GENERATOR_BASE_H_
+#define INCLUDE_COMPRESSIBLE_NAVIER_STOKES_POSTPROCESSOR_POINTWISE_OUTPUT_GENERATOR_BASE_H_
+
+#include <deal.II/base/conditional_ostream.h>
+#include <deal.II/base/hdf5.h>
+#include <deal.II/base/mpi_remote_point_evaluation.h>
+#include <deal.II/base/point.h>
+
+#include <deal.II/lac/la_parallel_vector.h>
+
+#include <deal.II/distributed/tria.h>
+
+#include <deal.II/fe/mapping.h>
+
+#include <deal.II/numerics/vector_tools.h>
+
+namespace ExaDG
+{
+template<int dim>
+struct PointwiseOutputDataBase
+{
+  using point_value_type = typename dealii::Point<dim>::value_type;
+
+  PointwiseOutputDataBase();
+
+  bool        write_output;
+  std::string directory;
+  std::string filename;
+
+  bool update_points_before_evaluation;
+
+  double start_time;
+  double end_time;
+  double interval_time;
+
+  std::vector<dealii::Point<dim>> evaluation_points;
+
+  void
+  print(dealii::ConditionalOStream & pcout) const;
+};
+
+template<int dim, typename Number>
+class PointwiseOutputGeneratorBase
+{
+public:
+  using VectorType       = dealii::LinearAlgebra::distributed::Vector<Number>;
+  using point_value_type = typename PointwiseOutputDataBase<dim>::point_value_type;
+
+  PointwiseOutputGeneratorBase(MPI_Comm const & comm);
+
+  virtual ~PointwiseOutputGeneratorBase() = default;
+
+  void
+  setup_base(dealii::DoFHandler<dim> const &      dof_handler_in,
+             dealii::Mapping<dim> const &         mapping_in,
+             PointwiseOutputDataBase<dim> const & pointwise_output_data_in);
+
+  void
+  evaluate(VectorType const & solution, double const & time, int const & time_step_number);
+
+  void
+  add_scalar_quantity_dataset(std::string const & name);
+
+  void
+  add_vector_quantity_datasets(std::string const & name);
+
+  void
+  write_scalar_quantity(std::string const & name, dealii::Vector<Number> const & hyperslab_data);
+
+  void
+  write_vectorial_quantity(std::string const & name, dealii::Vector<Number> const & hyperslab_data);
+
+  void
+  write_scalar_quantity(std::string const &         name,
+                        std::vector<Number> const & values,
+                        unsigned int const          first_component = 0)
+  {
+    AssertThrow(first_component == 0,
+                dealii::ExcMessage("This function can only be called with first_component==0."));
```

Comment:
also this looks a bit strange to me. Why is it needed in addition to the other function `write_scalar_quantity()` above?

Reply:
Remote Point evaluation gives us different types regarding the number of components that are present in the solution. Regarding the type, we have to extract values or not. With this function overload, the interface stays the same and the user does not have to care about it.

Reply:
Here, I am not convinced that this is the right design. We have a function argument that needs to be fixed to a certain value but is not used by the function.

Why not giving up scalar and vectorial completely? Why not using a single function with arguments `first_component` and `last_component` (maybe with the default value `last_component=first_component`) for the general case with `Tensor` as function argument. For the special case of a scalar quantity and `std::vector<Number>` as argument, wouldn't it make sense to write directly `write_component(name, std::move(dealii::Vector<Number>(values.begin(), values.end())))` instead of calling `write_scalar_quantity`?

Reply:
We need extract_component also in the scalar case since RPE gives a Tensor in which a scalar and a vector are combined.
Using first and last component would work

Reply:
To be consistent this would also require:
```add_quantity_dataset(std::string const&name,unsigned int const n_components=1)```
instead of ```add_scalar_quantity_dataset``` ```and add_vectorial_dataset```
Do you agree on this?

Reply:
I am not sure why there is `n_components` as argument. I would like to have only a single function like `add_dataset(name, values, first_component, last_component)`.

Reply:
Ah, I misunderstood it. You are right, here `n_components` is correct.

Reply:
Slightly different though. Hope you like it:
[2b20809](https://github.com/exadg/exadg/pull/255/commits/2b20809683371a883b4520bdc131134068aae0fd)

Full disclosure: I didn't test it. Before merging, I would like to ensure I didn't break anything by a stupid mistake.